Add pg_get_subscription_ddl() function

vaibhavdalvi93 opened the pull request:
This new SQL-callable function returns the `CREATE SUBSCRIPTION` statement for a given subscription name.

Like `pg_dump`, the returned DDL explicitly sets `connect = false`. This is because the original `CONNECT` option value is not cataloged, and using `connect = false` ensures the DDL can be successfully executed even if the remote publisher is unreachable.

By default, the function is restricted to superusers.  This is a security measure because subscription connection strings often contain sensitive information, such as passwords.  EXECUTE permission can be granted to other users as needed.

Author: Vaibhav Dalvi <vaibhav.dalvi@enterprisedb.com>
Discussion: https://www.postgresql.org/message-id/945db7c5-be75-45bf-b55b-cb1e56f2e3e9@dunslane.net

vaibhavdalvi93 committed:
Add pg_get_subscription_ddl() function

This new SQL-callable function returns the `CREATE SUBSCRIPTION`
statement for a given subscription name.

Like `pg_dump`, the returned DDL explicitly sets `connect = false`.
This is because the original `CONNECT` option value is not cataloged,
and using `connect = false` ensures the DDL can be successfully executed
even if the remote publisher is unreachable.

By default, the function is restricted to superusers.  This is a security
measure because subscription connection strings often contain sensitive
information, such as passwords.  EXECUTE permission can be granted to other
users as needed.

Author: Vaibhav Dalvi <vaibhav.dalvi@enterprisedb.com>
Discussion: https://www.postgresql.org/message-id/945db7c5-be75-45bf-b55b-cb1e56f2e3e9@dunslane.net

vaibhavdalvi93 commented:
@alvherre , Could you please review this PR.

ibarwick commented on src/backend/catalog/system_functions.sql:
```diff
@@ -782,6 +782,8 @@ REVOKE EXECUTE ON FUNCTION pg_ls_logicalmapdir() FROM PUBLIC;
 
 REVOKE EXECUTE ON FUNCTION pg_ls_replslotdir(text) FROM PUBLIC;
 
+REVOKE EXECUTE ON FUNCTION pg_get_subscription_ddl(text) FROM public;
```
We presumably don't want to do this, as the functions listed here are special cases?

ibarwick commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -13715,3 +13716,159 @@ get_range_partbound_string(List *bound_datums)
+	 * available in the catalog table.  We can use default values i.e. TRUE
+	 * for both.  This is what pg_dump uses.
+	 */
+	appendStringInfo(&buf, " WITH (connect = false");
```
Preferable to use `appendStringInfoString()` here as there are no parameters needing formatting (same for similar instances below).

ibarwick commented on src/include/catalog/pg_proc.dat:
```diff
@@ -3993,6 +3993,9 @@
 { oid => '1387', descr => 'constraint description',
   proname => 'pg_get_constraintdef', provolatile => 's', prorettype => 'text',
   proargtypes => 'oid', prosrc => 'pg_get_constraintdef' },
+{ oid => '703', descr => 'get CREATE statement for subscription',
```
I think normally the OID(s) used for patch submissions would be > 8000, as suggested by `src/include/catalog/unused_oids`.

drpalaric replied:
You should select an `oid` between the 8000-9999 range. See here: https://www.postgresql.org/docs/current/system-catalog-initial-data.html#SYSTEM-CATALOG-OID-ASSIGNMENT

> When choosing OIDs for a patch that is not expected to be committed immediately, best practice is to use a group of more-or-less consecutive OIDs starting with some random choice in the range 8000—9999.

Run:  `./src/include/catalog/unused_oids` and it will suggest one for you

akshay-joshi commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -13715,3 +13716,159 @@ get_range_partbound_string(List *bound_datums)
+	conninfo = TextDatumGetCString(datum);
+
+	/* Append connection info to the CREATE SUBSCRIPTION statement */
+	appendStringInfo(&buf, "CONNECTION \'%s\'", conninfo);
```
The password should not be displayed in plain text in the connection info string

akshay-joshi reviewed (changes requested):
There are 2 new parameters, **retain_dead_tuples** and **max_retention_duration**. I am not sure whether we need to add those to the DDL or not.

vaibhavdalvi93 committed:
Fixed 1st round of review comments

- use appendStringInfoString()
- use OID > 8000
- Add missed two options

gonzalemario commented on doc/src/sgml/func/func-info.sgml:
```diff
@@ -3797,4 +3797,54 @@ acl      | {postgres=arwdDxtm/postgres,foo=r/postgres}
+       </para>
+       <para>
+        Reconstructs the creating command for a subscription.
+        The result is a complete <command>CREATE SUBSCRIPTION</command>
```
Annoying nit, you have extra spaces after the dots

mw2q replied:
I thought it was still recommended to use two spaces after a period (end of sentence) as a visual aid (or ease of reading) for text?

gonzalemario replied:
interesting, I didn't know

gonzalemario commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -13715,3 +13716,159 @@ get_range_partbound_string(List *bound_datums)
+	datum = SysCacheGetAttrNotNull(SUBSCRIPTIONOID, tup,
+								   Anum_pg_subscription_subpublications);
+	publist = textarray_to_stringlist(DatumGetArrayTypeP(datum));
+	GetPublicationsStr(publist, pubnames, false);
```
Should `pubnames` be passed by reference?

drpalaric commented on doc/src/sgml/func/func-info.sgml:
```diff
@@ -3797,4 +3797,54 @@ acl      | {postgres=arwdDxtm/postgres,foo=r/postgres}
+       <para>
+        Reconstructs the creating command for a subscription.
+        The result is a complete <command>CREATE SUBSCRIPTION</command>
+        statement.  This statement omits default values.  The
```
This is another annoying nit. There should be 1 space after a full stop.

drpalaric commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -13715,3 +13716,159 @@ get_range_partbound_string(List *bound_datums)
+	Datum		datum;
+	bool		isnull;
+
+	if (PG_ARGISNULL(0))
```
Not sure you actually need this if you have `procisstrict => 't'` in your pg_proc.dat file: https://www.postgresql.org/docs/current/catalog-pg-proc.html

drpalaric commented on src/include/catalog/pg_proc.dat:
```diff
@@ -3993,6 +3993,9 @@
   proname => 'pg_get_constraintdef', provolatile => 's', prorettype => 'text',
   proargtypes => 'oid', prosrc => 'pg_get_constraintdef' },
+{ oid => '703', descr => 'get CREATE statement for subscription',
+  proname => 'pg_get_subscription_ddl', prorettype => 'text',
```
Here's where you can add `procisstrict => 't'`

vaibhavdalvi93 committed:
work on 2nd round of review comments

- Allow users to access this function those have
  pg_create_subscription and pg_read_all_data permission
- Don't omit default value because it may change in future

Author: Vaibhav Dalvi <vaibhav.dalvi@enterprisedb.com>
Reviewed-by: Nishant Sharma,
Reviewed-by: Ian Barwick
Reviewed-by: Akshay Joshi
Reviewed-by: Phil Alger
Discussion: https://www.postgresql.org/message-id/945db7c5-be75-45bf-b55b-cb1e56f2e3e9@dunslane.net

vaibhavdalvi93 committed:
fix check-world failure

The error ""wal_level" is insufficient to create the
replication slot required by retain_dead_tuples" was
there due to true "retain_dead_tuples".  Set to false.

Vaibhav Dalvi
